xtractmime/__init__.py

```diff
@@ -48,8 +69,33 @@ def extract_mime(
     extra_types: Optional[Tuple[Tuple[bytes, bytes, Set[bytes], str], ...]] = None,
 ) -> str:
     extra_types = extra_types or tuple()
-    # supplied_type = content_types[-1] if content_types else None
+    supplied_type = content_types[-1] if content_types else None
+    check_for_apache = http_origin and supplied_type in _APACHE_TYPES
+    resource_header = memoryview(body) if len(body) < 1445 else memoryview(body)[:1445]
+
+    if supplied_type in (None, "unknown/unknown", "application/unknown", "*/*"):
+        _find_unknown_mimetype(body, not no_sniff)
+
+    if no_sniff:
+        return supplied_type
+
+    if check_for_apache:
+        return _sniff_mislabled_binary(body)
+
+    if supplied_type[-4:] is "+xml" or supplied_type in ("text/xml", "application/xml"):
+        return supplied_type
+
+    if supplied_type is "text/html":
+        return _compare_feed_html(body)
+
+    matched_type = is_image(resource_header)
+
+    if matched_type:
+        return matched_type
```

Comment:
I believe this is missing the “If the supplied MIME type is an image MIME type supported by the user agent” part, and similar for audio and video later.

I’ve just found out about the concept of [MIME type supported by the user agent](https://mimesniff.spec.whatwg.org/#supported-by-the-user-agent), which seems to affect the image and audio/video logic here (we should only return the matched type if it is supported by the user agent). So I guess we should add one more parameter to `extract_mime` for a set of MIME types that the calling user agent supports. We could have the parameter be `None` by default, and interpret that as supporting any MIME type.

With that parameter, and taking into account “If the supplied MIME type is an image MIME type” as well, these lines could become something like:

```suggestion
    if supplied_type.startswith('image/'):
        matched_type = is_image(resource_header)
        if matched_type in supported_types:
            return matched_type
```

Also, `is_…` sounds like it is or returns a boolean value. Maybe something like `get_image_mime` would be better to name that and similar functions?

Reply:
> I believe this is missing the “If the supplied MIME type is an image MIME type supported by the user agent” part, and similar for audio and video later.

I think I misinterpret this part, I will fix it.
>Also, is_… sounds like it is or returns a boolean value. Maybe something like get_image_mime would be better to name that and similar functions?
yeah, this name `get_image_mime` is more suitable for the function. I will apply this change to the rest of the functions :+1: